File: sidecar/src/env/paths.ts
// Cross-platform paths for the bundled Python env.
//
// Three layers:
//   1. Resource dir  — read-only, ships in the installer.
//      Set by the Tauri Rust supervisor via BIOCLAW_RESOURCE_DIR.
//      Contains: `bioclaw-env/{pyproject.toml,uv.lock,.python-version,README.md}`
//      plus the bundled `uv` binary (next to the sidecar in
//      `binaries/uv-<rustc-triple>`).
//   2. Project dir   — writable, lives under the user's home.
//      Default: `~/.bioclaw/env/`. This is where we COPY the
//      resource-dir source files on first setup and where uv writes
//      `.venv/`. Survives app upgrades; users can wipe it to force a
//      re-sync.
//   3. uv data dir   — uv's own cache + downloaded interpreters,
//      shared across BioClaw versions (and any other uv consumer on
//      the machine). uv chooses this itself unless we override.

import path from 'node:path';
import os from 'node:os';
import process from 'node:process';

/** Default user-side project dir. Override via BIOCLAW_ENV_DIR. */
export function defaultProjectDir(): string {
  const override = process.env['BIOCLAW_ENV_DIR'];
  if (override && override.length > 0) return override;
  return path.join(os.homedir(), '.bioclaw', 'env');
}

/** Per-platform venv interpreter path inside the project dir. */
export function venvPython(projectDir: string): string {
  if (process.platform === 'win32') {
    return path.join(projectDir, '.venv', 'Scripts', 'python.exe');
  }
  return path.join(projectDir, '.venv', 'bin', 'python');
}

/** Tauri's resource_dir() lands here at runtime — set by Rust supervisor. */
export function resourceDir(): string | null {
  const v = process.env['BIOCLAW_RESOURCE_DIR'];
  return v && v.length > 0 ? v : null;
}

/** Where the bundled env sources live within resource_dir. */
export function bundledEnvSourceDir(): string | null {
  const r = resourceDir();
  return r ? path.join(r, 'bioclaw-env') : null;
}

/** Pre-baked env zip (CPython + uv-cache + sources) — primary install
 *  path. Tauri ships it as a bundle resource at
 *  `<resource_dir>/bioclaw-env.zip`. The legacy
 *  `bundledEnvSourceDir()` (pyproject + lock as loose files) is still
 *  accepted as a fallback for the network-install code path. */
export function bundledEnvZip(): string | null {
  const r = resourceDir();
  if (!r) return null;
  const p = path.join(r, 'bioclaw-env.zip');
  return p;
}

/** Where the bundled uv binary lives. tauri-bundler renames
 *  `binaries/uv-<rustc-triple>(.exe)` to plain `uv(.exe)` at install
 *  time and drops it next to the main binary. We probe both the
 *  resource dir (Linux .deb layout) and the install dir + PATH for
 *  resilience. */
export function bundledUvBinary(): string {
  return process.platform === 'win32' ? 'uv.exe' : 'uv';
}

File: sidecar-rs/src/env/paths.rs
//! Cross-platform path resolution for the bundled Python env.
//!
//! Three layers (same as the Node port):
//! 1. Resource dir — read-only, ships in the installer. Tauri sets
//!    `BIOCLAW_RESOURCE_DIR`; `bioclaw-env.zip` + `bioclaw-env/`
//!    source files live under here.
//! 2. Project dir — writable, under the user's home. Defaults to
//!    `~/.bioclaw/env`; `--env-dir` / `BIOCLAW_ENV_DIR` overrides.
//! 3. uv data dir — uv's own cache + downloaded interpreters,
//!    shared across BioClaw versions. uv picks this itself unless
//!    we pin it via `UV_PYTHON_INSTALL_DIR` for the offline path.

use std::path::{Path, PathBuf};

/// Default project dir. Mirrors `sidecar/src/env/paths.ts::defaultProjectDir`
/// — POSIX uses `~/.bioclaw/env` (preserving the Node sidecar's layout
/// so existing users' envs don't move); Windows uses
/// `%LOCALAPPDATA%\BioClaw\env`.
pub fn default_project_dir() -> Option<PathBuf> {
    if let Ok(v) = std::env::var("BIOCLAW_ENV_DIR") {
        if !v.is_empty() {
            return Some(PathBuf::from(v));
        }
    }
    if cfg!(target_os = "windows") {
        let dirs = directories::ProjectDirs::from("tech", "bioclaw", "BioClaw")?;
        Some(dirs.data_local_dir().join("env"))
    } else {
        let home = std::env::var_os("HOME")?;
        let mut p = PathBuf::from(home);
        p.push(".bioclaw");
        p.push("env");
        Some(p)
    }
}

/// Per-platform venv interpreter path inside the project dir.
/// `.venv/Scripts/python.exe` on Windows, `.venv/bin/python` elsewhere.
pub fn venv_python(project_dir: &Path) -> PathBuf {
    if cfg!(target_os = "windows") {
        project_dir.join(".venv").join("Scripts").join("python.exe")
    } else {
        project_dir.join(".venv").join("bin").join("python")
    }
}

/// Bundled env source dir under resource_dir. Returns None when no
/// resource dir is configured.
pub fn bundled_env_source_dir(resource_dir: Option<&Path>) -> Option<PathBuf> {
    resource_dir.map(|r| r.join("bioclaw-env"))
}

/// Pre-baked env zip — primary install path. Tauri ships it as a
/// bundle resource at `<resource_dir>/bioclaw-env.zip`.
pub fn bundled_env_zip(resource_dir: Option<&Path>) -> Option<PathBuf> {
    resource_dir.map(|r| r.join("bioclaw-env.zip"))
}

/// Name of the bundled uv binary (extension differs on Windows).
pub fn bundled_uv_binary_name() -> &'static str {
    if cfg!(target_os = "windows") {
        "uv.exe"
    } else {
        "uv"
    }
}
